src/System.Text.JsonLab/System/Text/Json/JsonReader.cs

```diff
@@ -161,24 +160,19 @@ private bool ReadMultiSegment()
 
         private bool ReadSingleSegment()
         {
-            ref byte bytes = ref MemoryMarshal.GetReference(_buffer);
-            int length = _buffer.Length;
-            int skip = SkipWhiteSpaceUtf8(ref bytes, length);
+            SkipWhiteSpaceUtf8();
 
-            ref byte next = ref Unsafe.Add(ref bytes, skip);
-            length -= skip;
-
-            if (length < 1)
+            if (_buffer.Length < 1)
```

Comment:
`(uint)_buffer.Length < 1` help skip bounds check?

Reply:
I don't see any difference in disassembly. Which bounds check would get skipped?

![image](https://user-images.githubusercontent.com/6527137/43553962-7cbbbbe0-95a6-11e8-9e2f-11a77bbbe641.png)

Reply:
The access `byte ch = _buffer[0];`
```
cmp ebx, 0
jbe 00007FFB7...
```
Same with `Unsafe.Add(ref src, 0)` that have become `_buffer[0]` etc

Try greater than?
```csharp
if (1u > (uint)_buffer.Length)
{
    return false;
}
byte ch = buffer[0];
```
Otherwise perhaps it needs to be a local rather than field access?
```csharp
var buffer = _buffer;
if (1u > (uint)_buffer.Length)
{
    return false;
}
byte ch = buffer[0];
```

But the Jit doesn't remove the bounds check if its not `uint` and doesn't work that out even if its a constant https://github.com/dotnet/coreclr/issues/18934

Reply:
**Casting to uint or changing to greater than has no affect.**

Copying the span into a local gets rid of the extra bounds checks. However, if there is only one access, then the difference is between a struct copy vs. a bounds check (a wash).

https://sharplab.io/#v2:C4LghgzgtgPgrhAlgOwOYAIDKBPCwCmUA3ALABQADnAEYA2iAxugE74Bm6ezcDw6AwugDe5dGPQVmiAG5gC6AEr4wAEwDyyWtkwUwyADzVsBAHzoA+tThs2+ZqTLj0o8VTqN01APZfa6AHIAjAAUAJTCLk5iiBzBlta2zAB0ADL4aMAAFuj66ADM6DAwFlY2dgDaAAwAuugAvHXogYXF8WXM5YG1DegATC0lCRW93Y15oZFRIo5Rs6zAcMzI6GxgtBD4DrNiAL6TTvOLy8Dcm5N7M2Ju9EzevgG9YRGXUTHowcFwKMChbYmp6VQWRy+TgAz+FRq9UazSKg3anVGfXBpUS5RG0PyExeTmm2yihyWKzWGy22wu+MJx1OZN2k2uHjufn8eSeeO2SlUGi0Oj0hmM+DMqLs9XhiVpr1iwuSaQy2VyBTh0qqSNhxWVXUx/SVQw6GJ64324nZ+LEVOJ6zOOPEFO25pOcCtTltVxoN08PmZABY2UborE8mCzJ9vr9pQC5SjdSrMWqxRVNT1ta1lfqxtj8SbTebVpaJc6/Sx8AsiQ6nTbyDsgA===

If we have multiple span accesses, then it helps.
![image](https://user-images.githubusercontent.com/6527137/43555703-b724873a-95b0-11e8-85dc-f386ab5fdacd.png)

Reply:
> Copying the span into a local gets rid of the extra bounds checks. However, if there is only one access, then the difference is between a struct copy vs. a bounds check (a wash).

Its a shame the copy to local is needed to elide the bounds check. It makes sense with an array as the field reference may be changed to a different one by another thread (I'm assuming that's why its conservative); whereas the span is stack local so another thread won't be changing it so the Jit doesn't have to be as conservative? Might be completely wrong about this though... /cc @AndyAyersMS @mikedn

Reply:
Yes, when a field comes from a byref-like type then the jit should be able to assume the field is not cross-thread visible and not be quite so conservative.

Reply:
AFAIR I looked at a similar issue recently. Being stack only does not mean that other threads can't touch it, they may receive the address of the local by some means (e.g. via a static pointer typed variable). However, this seems impossible to do in a verifiable manner so the JIT could probably be more liberal with ref struct field accesses.